fdbserver/DataDistributionQueue.actor.cpp

```diff
@@ -1765,35 +1917,34 @@ ACTOR Future<bool> rebalanceReadLoad(DDQueueData* self,
 		return false;
 	}
 	// check team difference
-	auto srcLoad = sourceTeam->getLoadReadBandwidth(false), destLoad = destTeam->getLoadReadBandwidth();
+	state double srcLoad = sourceTeam->getLoadReadBandwidth(false);
+	state double destLoad = destTeam->getLoadReadBandwidth();
 	traceEvent->detail("SrcReadBandwidth", srcLoad).detail("DestReadBandwidth", destLoad);
 
 	// read bandwidth difference is less than 30% of src load
 	if ((1.0 - SERVER_KNOBS->READ_REBALANCE_DIFF_FRAC) * srcLoad <= destLoad) {
 		traceEvent->detail("SkipReason", "TeamTooSimilar");
 		return false;
 	}
-	// randomly choose topK shards
-	int topK = std::min(int(0.1 * shards.size()), SERVER_KNOBS->READ_REBALANCE_SHARD_TOPK);
+
 	state Future<HealthMetrics> healthMetrics = self->cx->getHealthMetrics(true);
-	state GetTopKMetricsRequest req(
-	    shards, topK, (srcLoad - destLoad) * SERVER_KNOBS->READ_REBALANCE_MAX_SHARD_FRAC, srcLoad / shards.size());
-	req.comparator = [](const StorageMetrics& a, const StorageMetrics& b) {
-		return a.bytesReadPerKSecond / std::max(a.bytes * 1.0, 1.0) >
-		       b.bytesReadPerKSecond / std::max(b.bytes * 1.0, 1.0);
-	};
-	state GetTopKMetricsReply reply = wait(brokenPromiseToNever(self->getTopKMetrics.getReply(req)));
 	wait(ready(healthMetrics));
-	auto cpu = getWorstCpu(healthMetrics.get(), sourceTeam->getServerIDs());
+	auto cpu = getWorstCpu(healthMetrics.get(), sourceTeam->getServerIDs()); // Server cpu
 	if (cpu < SERVER_KNOBS->READ_REBALANCE_CPU_THRESHOLD) { // 15.0 +- (0.3 * 15) < 20.0
 		traceEvent->detail("SkipReason", "LowReadLoad").detail("WorstSrcCpu", cpu);
 		return false;
 	}
 
-	auto& metricsList = reply.metrics;
+	// randomly choose topK shards
+	state GetTopKMetricsReply topKReply = wait(GetTopOrBottomKMetrics(self, shards, true, srcLoad, destLoad));
```

Comment:
We should one or the other, not both.
Currently, it would be based on whether the `reason` is `MountainChopper` or `ValleyFiller`. Yes?

Reply:
MountainChopper work: `reduceShards` + `multiplyShards` + `moveShards`
ValleyFiller work:  `reduceShards` + `moveShards`